ports/esp32/machine_encoder.c

```diff
@@ -0,0 +1,787 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+/*
+ESP32 Pulse Counter
+https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/pcnt.html
+Wrapped around
+https://github.com/espressif/esp-idf/blob/master/components/driver/include/driver/pcnt.h
+https://github.com/espressif/esp-idf/blob/master/components/hal/include/hal/pcnt_types.h
+https://github.com/espressif/esp-idf/blob/master/components/driver/pcnt.c
+See also
+https://github.com/espressif/esp-idf/tree/master/examples/peripherals/pcnt/pulse_count_event
+*/
+
+/*
+ESP32 Quadrature Counter based on Pulse Counter(PCNT)
+Based on
+https://github.com/madhephaestus/ESP32Encoder
+https://github.com/bboser/MicroPython_ESP32_psRAM_LoBo/blob/quad_decoder/MicroPython_BUILD/components/micropython/esp32/machine_dec.c
+See also
+https://github.com/espressif/esp-idf/tree/master/examples/peripherals/pcnt/rotary_encoder
+*/
+
+#include "py/runtime.h"
+#include "mphalport.h"
+#include "modmachine.h"
+
+#if MICROPY_PY_MACHINE_PCNT
+
+#include "driver/pcnt.h"
+#include "soc/pcnt_struct.h"
+#include "esp_err.h"
+
+#include "machine_encoder.h"
+
+#define GET_INT mp_obj_get_int_truncated
+// #define GET_INT mp_obj_get_ll_int // need PR: py\obj.c: Get 64-bit integer arg. #80896
+
+STATIC pcnt_isr_handle_t pcnt_isr_handle = NULL;
+STATIC mp_pcnt_obj_t *pcnts[PCNT_UNIT_MAX] = {};
+
+#if ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(4, 1, 0)
+#define EVT_THRES_0 PCNT_EVT_THRES_0
+#define EVT_THRES_1 PCNT_EVT_THRES_1
+#define EVT_ZERO    PCNT_EVT_ZERO
+#else
+#define EVT_THRES_0 (1 << PCNT_EVT_THRES_0)
+#define EVT_THRES_1 (1 << PCNT_EVT_THRES_1)
+#define EVT_ZERO    (1 << PCNT_EVT_ZERO)
+#endif
+
+/* Decode what PCNT's unit originated an interrupt
+ * and pass this information together with the event type
+ * the main program using a queue.
+ */
+#if CONFIG_IDF_TARGET_ESP32S2 || CONFIG_IDF_TARGET_ESP32S3
+#define H_LIM_LAT  cnt_thr_h_lim_lat_un
+#define L_LIM_LAT  cnt_thr_l_lim_lat_un
+#define THRES0_LAT cnt_thr_thres0_lat_un
+#define THRES1_LAT cnt_thr_thres0_lat_un
+#define ZERO_LAT   cnt_thr_zero_lat_un
+#else
+#define H_LIM_LAT  h_lim_lat
+#define L_LIM_LAT  l_lim_lat
+#define THRES0_LAT thres0_lat
+#define THRES1_LAT thres1_lat
+#define ZERO_LAT   zero_lat
+#endif
+STATIC void IRAM_ATTR pcnt_intr_handler(void *arg) {
+    for (int id = 0; id < PCNT_UNIT_MAX; ++id) {
+        if (PCNT.int_st.val & (1 << id)) {
+            mp_pcnt_obj_t *self = pcnts[id];
+            if (self != NULL) {
+                if (PCNT.status_unit[id].H_LIM_LAT) {
```

Comment:
This has changed in IDF 4.3 and there's now [pcnt_get_event_status()](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/pcnt.html#_CPPv421pcnt_get_event_status11pcnt_unit_tP8uint32_t)) for this